Bump schema to v3

Shrews opened the pull request:
Revamps the code and documentation to:

- Define v2 schema to match what we released in builder `1.2` (image validation support).
- Define a new v3 schema that adds a boat load of new features, and removes cruft we no longer want to support (`ansible_config` and builder images).

Shrews committed:
Bump schema version 3

Shrews committed:
Name final image

nitzmahone commented on docs/definition.rst:
```diff
@@ -17,7 +18,8 @@ If the EE file does not specify a version, version 1 will be assumed.
 ----------------
 
-An example version 1 execution environment definition schema is as follows:
+Example V1
```
After reading down this whole thing from the perspective of someone new to builder, I think we might want a standalone "from scratch" description of the current schema with examples, then have the descriptions of the old schemas and the deltas below it. That way, new users can be :see_no_evil: about the old stuff and just have a one-stop-shop for "how do I write this?" (and also not be tripped up by stopping reading before getting to the good stuff), while the reference delta descriptions and examples for the old schemas are still available below for those that need it. We should probably also discuss timing for deprecating v1/v2 schemas (ie, do we do it in this release, next release, something else?)

I've also been half-tempted to just call this one builder 3.0 to avoid the inherent confusion of the schema version drifting from the release version...

Shrews replied:
Reduced schema doc to v3 only.

nitzmahone commented on docs/definition.rst:
```diff
@@ -129,34 +133,29 @@ section, either for before the main build steps (``prepend``) or after
+    below) to control Podman image validation, it is up to the user to properly
+    configure the Podman runtime to talk to the registries needed. This may include
+    defining the sigstore for each registry, using secure connections (or not), etc.
+    Such configuration is beyond the scope of this document.
```
+1000 :wink:

nitzmahone commented on ansible_builder/ee_schema.py:
```diff
@@ -228,14 +308,16 @@ def validate_schema(ee_def: dict):
             raise DefinitionError(f"Schema version not an integer: {ee_def['version']}")
 
-    if schema_version not in (1, 2):
+    if schema_version not in (1, 2, 3):
```
It might be more trouble than it's worth, but I also wonder if it'd be sane to do a simple version divination heuristic to not require an explicit version. The developer in me wants people to be specific, but the user in me is lazy... :laughing: Maybe something like:

* v2 requires explicit opt-in (since IIRC it always did anyway)
* if we don't see a version:
  * check for presence of v1 keys, force v1 mode if present (which will fail if any v2/v3 keys are in use, good)
  * otherwise assume latest; if we ever make other breaking changes in future schema versions, we'll figure something out (and worst case go back to requiring a schema version again for newer versions)

Shrews replied:
Yeah, I don't know what to do with this one, really. Technically, an EE can be just a `dependencies` section with, say, only python requirements. That's common across all versions, atm. Which do we choose then? I think this would require quite a bit more thought than we have time for right now.

nitzmahone replied:
Fair enough- I hadn't considered the "absolutely minimalist" cases where v1/v3 are indistinguishable...

nitzmahone reviewed (commented):
Looking good- a couple of philosophical questions to kick around before deciding the next step...

Shrews committed:
Update definition doc for v3 and fix v3 test

sivel commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
-      python: requirements.txt
+      python:
+        - six
+        - psutil
```
The `schema_v3` doesn't seem to agree with this documentation.  

```
                "python": {
                    "description": "The python dependency file",
                    "type": "string",
                },
```

sivel replied:
FWIW, I like what the docs have, as it reduces the dependencies on external files.  But if we did extend the schema to support this, I think we should do it for `galaxy` and `bindep` too.

Shrews replied:
Heh. Neat.

Support for inline values was added for all three keys (`python`, `galaxy`, and `bindep`) in @nitzmahone's original PR. I added jsonschema validation after that and apparently didn't adjust `python` or `bindep` to support lists (`galaxy` must still be a string, but can be the YAML contents, not just a filename). But oddly, it somehow works. The code that handles it is here:

https://github.com/nitzmahone/ansible-builder/blob/648d4f705d7ae6590b7fa76c41d874024147a323/ansible_builder/user_definition.py#L164-L175

Current guess is that the inline list comes in as a string but somehow magic happens. Will explore why this is working as-is in the morning. But yes, we should fix the schema too.

Shrews replied:
So, this wasn't actually working. Must've broke it when jsonschema support was added. Fixed with tests.

Shrews committed:
Various v3 bug fixes

- Fix the v3 schema for inline values and add tests.
- Always run ensure pip in base image so pip is available for later stages.
- Set UserDefinition.builder_image only when a builder image is specified
  so checks for it work correctly.

nitzmahone commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
-signed container images. This feature is only supported with the ``podman``
-container runtime.
+  * Version 2: Supported by ``ansible-builder`` versions ``1.2`` and later.
+  * Version 3: Supported by ``ansible-builder`` versions after ``1.2``.
```
I assume we'll fill in an actual version number once we finalize that choice...

Shrews replied:
Sure, we can at some point. I changed it to this (instead of referencing `2.0`) because you said you were tempted to move the version to 3.0 to match the schema version.

nitzmahone commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
-This integer value defines the version of the EE file. If not specified, the
-default of ``1`` will be used.
+This section allows you to add any file to the build context directory. These can
+then be referenced at any of image build stages. The format is a list of dictionary
```
Since folks using this tool aren't necessarily container-savvy, might want to clarify that this doesn't actually copy files into the final image- something like:

```suggestion
then be referenced or copied by `additional_build_steps` during any build stage. The format is a list of dictionary
```

nitzmahone commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
+      may either be an absolute path (e.g., ``/home/user/.ansible.cfg``),
+      or a path that is relative to the execution environment file. Relative paths may be
+      a regular expression matching one or more files (e.g. ``files/*.cfg``). Note
+      that the absolute path may *not* include a regular expression. If ``src`` is
```
s/the/an/

nitzmahone commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
+      Specifies the source file(s) to copy into the build context directory. This
+      may either be an absolute path (e.g., ``/home/user/.ansible.cfg``),
+      or a path that is relative to the execution environment file. Relative paths may be
+      a regular expression matching one or more files (e.g. ``files/*.cfg``). Note
```
Should we use the more accurate term "glob" or "glob expression" in this section?

nitzmahone commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
 
-``ANSIBLE_GALAXY_CLI_ROLE_OPTS``
-  This allows the user to pass the flags to the Role installation.
+This section enables you to specify custom build commands in for any build phase.
```
Maybe also add in here that the lines they're adding are `Containerfile` / `Dockerfile` directives, eg `RUN`, `ADD`, `FROM`, whatever, that need to conform to the rules for whatever actual image builder they're using (`podman build`, `docker build`, Docker's `buildkit`, etc)

nitzmahone commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
-  The supplied value may be a relative path from the directory of the execution
-  environment definition's folder, or an absolute path.
+    ``ANSIBLE_GALAXY_CLI_COLLECTION_OPTS``
+      This allows the user to pass the '--pre' flag (or others) to enable the installation of pre-releases collections.
```
Maybe similar wording to the one below for role opts, then follow with "a common example is ..." and the `--pre` thing?

nitzmahone commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
-builder will build the resulting image, based on the value of the
-:ref:`container-policy` CLI option.
+    ``ansible_core``
+      The version of the Ansible python package to be installed by pip.
```
Should probably mention that whatever's here for both `ansible_core` will get passed directly to `pip install`, so it can be a local/http/git path, an arbitrary package name with or without a version, etc, and maybe some examples, eg:
`ansible_core: ansible-core` <-- latest from PyPI (or whatever the configured pip index is on the builder host)
`ansible_core: ansible-core==2.14.2` <-- exact ansible_core version from PyPI/configured index
`ansible_core: https://github.com/nitzmahone/ansible/archive/refs/heads/tempfile_secontext.tar.gz` <-- install from a GitHub PR tarball

nitzmahone commented on docs/definition.rst:
```diff
@@ -4,210 +4,224 @@ Execution Environment Definition
 
-.. note::
+    ``ansible_runner``
+      The version of the Ansible Runner python package to be installed by pip.
```
same deal as `ansible_core` above

nitzmahone reviewed (commented):
docs are looking great- a couple of suggested clarifications

Shrews committed:
Changes to definition.rst doc
